packages/codec/README.md

```diff
@@ -1,15 +1,161 @@
+these functions to ensure you get an ABI mode result.  As noted above, there's
+no way to ensure you get a full mode result.
+
+(There are two slight differences between running the decoder in full mode and
+abifying afterward, versus simply running the decoder in ABI mode.  Firstly,
+full mode will reject certain invalid decodings that ABI mode cannot recognize
+as invalid.  Secondly, the abified version of a full-mode decoding does contain
+slightly more information than an actual ABI-mode decoding, just in an
+ABI-mode-compatible format.)
+
+Note that modes are always applied at the level of the whole decoding; different
+variables in the same decoding will always be decoded with the same mode.
+However, if an object (such as a log) admits *multiple* decodings, these
+different decodings may occur in different modes.
+
+Note that decoding of state variables is only available in full mode; attempting
+to decode state variables will result in an exception if full mode is not
+possible.
+
+### Output format: `Type`s and `Result`s
+
+This will be a brief explanation of the format of decoded variables.  For full
+detail, I recommend seeing the documentation for `Format`.
+
+Each decoded value is a `Result`.  A `Result` has the following fields:
+
+1. `type`: This is a `Type` object describing the value's type.  Each `Type`
+has a `typeClass` field describing the overall broad type, such as `"uint"` or
+`"bytes"`, together with additional information that gives the specific type.  I
+won't go into further detail about these here; I recommend simply looking at
+`Format.Types` to see how these work.
+
+2. `kind`: This is either `"value"`, in which case the `Result` is a `Value`, or
+`"error"`, in which case the `Result` is an `ErrorResult`.  In the former case,
+there will be a `value` field containing the decoded value.  In the latter case,
+there will be an `error` field indicating what went wrong.  *Warning*: When
+decoding a complex type, such as an array, mapping, or array, getting a kind of
+`"value"` does not necessarily mean the individual elements were decoded
+successfully.  Even if the `Result` for the array (mapping, struct) as a whole
+has kind "value"`, the elements might still have kind `"error"`.
+
+3. `value`: As mentioned, this is included when `kind` is equal to `"value"`.
+It contains information about the actual decoded value.  I recommend seeing
+`Format.Values` for more information.
+
+4. `error`: The alternative to `value`.  Generally includes information about
+the raw data that led to the error.  See `Format.Errors` for more
+information.
+
+5. `reference`: This field is a debugger-only feature so I'll skip explaining it
+here; it won't come up when using this interface.
+ 
+Rather than explain the format in detail -- you can see the TypeDoc for that -- I'd
+instead like to take a moment to answer the question: What counts as a value,
+and what counts as an error?
```

Comment:
This paragraph maybe can just be replaced (or preceded by) a heading "Values vs. Errors"

Reply:
Hm, a subheading maybe.  Would need to find a way to make it flow.